### PR TITLE
Add GLM-5 Next support

### PR DESCRIPTION
**Pull request by Blaizzy**
## Summary

- add GLM-5 Next text and multimodal model support with a native processor
- generalize FP8 handling and optimize sparse prefill and decode
- add MTP speculative decoding support
- align configuration, preprocessing, sparse attention, vision, KDA masking, tied embeddings, video metadata, and frame sampling with Transformers 5.16.1

## Validation

- GLM-5 test suite: 580 passed, 53 subtests passed
- sparse scorer recall: 100% at 4K, 16K, and 64K contexts
- dense and MoE parity max error: 2.38e-7
- vision parity max error: 1.24e-10
- KDA valid-token max error: 1.34e-7; padding behavior matches exactly
- video frame sampling parity: 1000/1000 cases
- checkpoint configuration parity: 59/59 fields
- Black, isort, autoflake, and diff checks pass

### CODE REVIEW

**Comment by Lazarus-931** [comment]
nice!

can u also fold https://github.com/Blaizzy/mlx-vlm/pull/2074 which has dflash2 and I can close that one!

**Comment by Blaizzy** [comment]
Sure, will do @Lazarus-931

**Comment by Blaizzy** [comment]
### Direct A/B against #2044

I ran this branch against #2044 at head `91b51a3` on the same M3 Ultra (512 GB) and the same checkpoints.

Setup:

- GLM-5.3-Flash 4-bit target + FP8 MTP drafter
- greedy decoding, fixed MTP block size 2
- 200 generated tokens per row
- exact prompt contexts: 512, 2,048, 4,096, 8,192, and 16,384 tokens
- batch sizes 1, 2, and 4, with identical prompts within each batch
- shape warmups before measurement

This branch serializes several generalized projections in fused output-row form, whereas #2044 expects separate module names. For the #2044 run I used a lazy, lossless load adapter that only slices those packed output rows and remaps keys. Strict validation matched 2,651/2,651 target tensors and 39/39 drafter tensors, with no missing/unexpected keys, dequantization, requantization, or changes to #2044's compute path.

| Batch | Context | This base | #2044 base | This MTP | #2044 MTP | This boost | #2044 boost | MTP winner |
|---:|---:|---:|---:|---:|---:|---:|---:|:---|
| 1 | 512 | 24.19 | 30.85 | 37.48 | 41.65 | 1.550x | 1.350x | #2044 |
| 1 | 2,048 | 23.67 | 27.03 | 33.56 | 30.66 | 1.418x | 1.134x | this branch |
| 1 | 4,096 | 23.68 | 26.98 | 31.54 | 30.29 | 1.332x | 1.123x | this branch |
| 1 | 8,192 | 23.58 | 26.77 | 29.36 | 29.47 | 1.245x | 1.101x | #2044 |
| 1 | 16,384 | 23.43 | 26.51 | 31.30 | 27.90 | 1.336x | 1.052x | this branch |
| 2 | 512 | 39.88 | 47.42 | 51.69 | 57.74 | 1.296x | 1.217x | #2044 |
| 2 | 2,048 | 38.85 | 41.62 | 48.43 | 43.69 | 1.247x | 1.050x | this branch |
| 2 | 4,096 | 38.59 | 41.36 | 46.77 | 44.34 | 1.212x | 1.072x | this branch |
| 2 | 8,192 | 38.28 | 41.24 | 45.35 | 42.66 | 1.185x | 1.035x | this branch |
| 2 | 16,384 | 42.30 | 40.28 | 50.06 | 43.37 | 1.183x | 1.077x | this branch |
| 4 | 512 | 69.52 | 76.48 | 90.53 | 91.78 | 1.302x | 1.200x | #2044 |
| 4 | 2,048 | 68.99 | 63.94 | 77.33 | 68.44 | 1.121x | 1.070x | this branch |
| 4 | 4,096 | 68.53 | 63.82 | 75.57 | 67.41 | 1.103x | 1.056x | this branch |
| 4 | 8,192 | 67.85 | 63.17 | 72.04 | 66.06 | 1.062x | 1.046x | this branch |
| 4 | 16,384 | 66.96 | 60.50 | 71.60 | 67.14 | 1.069x | 1.110x | this branch |

Summary:

- This branch wins MTP throughput in 11/15 cells.
- Output-token-weighted aggregate MTP throughput is 56.34 tok/s here versus 53.07 tok/s on #2044: +6.16%.
- Weighted aggregate base throughput is 46.27 tok/s here versus 48.26 tok/s on #2044.
- Aggregate base-to-MTP gain is 1.218x here versus 1.100x on #2044.
- Mean acceptance is 86.76% here versus 82.65% on #2044.
- Greedy base/MTP text parity is 15/15 here versus 1/15 on #2044 for this pure-4-bit target.

The #2044 generations remained locally coherent, but 14/15 MTP rows diverged from their own greedy base generation. A focused 512-token-context check diverged after 145 decoded characters. The most direct suspect is the quantized verifier projection: #2044's `verify_logits` uses singleton-exact projection for dense heads but falls back to the ordinary block projection for a quantized LM head; this branch has an exact quantized block argmax path.

These are same-machine/same-checkpoint results. They do not reproduce #2044's later posted 1.46-1.56x batch-1 table under this 200-token harness.

**Comment by Blaizzy** [comment]
Implemented the batch-wide verifier optimizations.

| Batch | Base tok/s | MTP tok/s | Speedup | Acceptance | Parity |
|---:|---:|---:|---:|---:|:---:|
| 1 | 29.48 | 43.67 | 1.481× | 90.5% | Yes |
| 2 | 45.09 | 64.54 | 1.431× | 85.7% | Yes |
| 4 | 70.77 | 83.03 | 1.173× | 81.8% | Yes |
| 8 | 104.10 | 91.38 | 0.878× | 48.1% | Yes |
| 16 | 127.86 | 140.76 | 1.101× | 81.8% | Yes |
| 32* | 159.32 | 128.40 | 0.806× | 46.2% | Yes |

*Batch 32 used 20 output tokens; the other rows used 40. Context was 512 and block size was 2.*

Notably, the previously regressed results recovered:

- Batch 1: 31.16 → 43.67 MTP tok/s
- Batch 4: 68.71 → 83.03 MTP tok/s

Changes include:

- Runtime batch-wide quantized dispatch in [quantized_verifier.py](https://github.com/Blaizzy/mlx-vlm/blob/pc/add-glm5-next/mlx_vlm/models/quantized_verifier.py).
- Batch-grouped MoE expert execution in [speculative_verifier.py](https://github.com/Blaizzy/mlx-vlm/blob/pc/add-glm5-next/mlx_vlm/models/glm5_next/speculative_verifier.py).
- Restored fused batch-1 hyperconnection ingress.
- Captured GDN states for rollback, cutting rejection rollback from about 30.0 ms to 14.7 ms.

There is no software batch ceiling; quantized parity tests cover batches through 64 and affine 2/3/4/5/6/8-bit, MXFP4, MXFP8, and NVFP4.

All 534 speculative tests and pre-commit checks pass. The batch-8/32 regressions are now acceptance-limited, not kernel-limited: below 50% acceptance, MTP cannot amortize verification and rollback work.

Implemented in commit [`9c5fa04e`](https://github.com/Blaizzy/mlx-vlm/commit/9c5fa04e).

File: mlx_vlm/models/sparse_attention.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
+from functools import lru_cache
```
Comment (Lazarus-931) [inline mlx_vlm/models/sparse_attention.py]: for clarity, would it be better if this was in kernels/?

File: mlx_vlm/models/quantized_verifier.py [hunk 2]
```diff
@@ -0,0 +1,1485 @@
+"""Format-independent quantized operations for exact speculative verification."""
```
Comment (Lazarus-931) [inline mlx_vlm/models/quantized_verifier.py]: and this in speculative/?
